### PR TITLE
Enable server side cursors

### LINKED ISSUE #35689: ENH: Support PostgreSQL server-side cursors to prevent memory hog on large datasets
#### Is your feature request related to a problem?

`pandas.read_sql_query` supports Python "generator" pattern when providing `chunksize` argument. It's not very helpful when working with large datasets, since the whole data is initially retrieved from DB into client-side memory and later chunked into separate frames based on `chunksize`. Large datasets will easily run into out-of-memory problems with this approach.

#### Describe the solution you'd like

Postgres/psycopg2 are addressing this problem with [server-side cursors](https://www.psycopg.org/docs/usage.html#server-side-cursors). But Pandas does not support it.

#### API breaking implications

`is_cursor` argument of `SQLDatabase` or `SQLiteDatabase` is not exposed in `pandas.read_sql_query` or `pandas.read_sql_table` without any reason. It should be exposed, this way named (server-side) cursor could be provided.

#### Describe alternatives you've considered

Instead of doing:

```
iter = sql.read_sql_query(sql,
      conn,
      index_col='col1',
      chunksize=chunksize)
```

I tried reimplementing it like this:

```
from pandas.io.sql import SQLiteDatabase

curs = conn.cursor(name='cur_name') # server side cursor creation
curs.itersize = chunksize

pandas_sql = SQLiteDatabase(curs, is_cursor=True)
iter = pandas_sql.read_query(
      sql,
      index_col='col1',
      chunksize=chunksize)
```

but it fails because `SQLiteDatabase` tries to access `cursor.description`, which is NULL for some reason with server-side cursors (and idea why?).

### PR DESCRIPTION
**Pull request by J0**
This pull request attempts to fix #35689. I read an article by @itamarst and decided to look further into the codebase. After noticing that the fix semed to be simple I decided file a PR. The fix is naive and I'm not sure the true solution is as simple as I imagine it to be. There are probably many things I'm missing as I've not read the codebase in detail so do let me know what else needs to be addressed.

- [x] closes #35689
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry(N.A.)

### CODE REVIEW

**Review by auvipy — APPROVED** [review]

**Comment by itamarst** [comment]
Thanks for doing this! You could perhaps only do the stream_results=True conditionally, if chunksize is set.

**Review by jreback — CHANGES_REQUESTED** [review]
this doesnt' seems to break anything, but its certainly possible we are not exerciing things. can you a) comment on what that parameter is doing (e.g. put an reference in the code to the article), and b) if you can come up with a test that fails current and passes with this and c) would need a whatsnew note (perf section)? for 1.3.

would we ever need to turn this off?

**Comment by itamarst** [comment]
As context, this is a memory usage performance improvement (in some cases). You can see the article here (https://pythonspeed.com/articles/pandas-sql-chunking/) with some memory-usage graphs.

With streaming mode, rows are only fetched on demand. If you're using chunking, this is what you want, because otherwise you might try to load gigabytes of rows into memory and then dole them out in chunks, potentially running out of memory.

If you're _not_ using chunking, though, streaming is probably a runtime performance hint, because there's a bunch more network latency involved, and you already know you want all the data so why not just get it all in one go? (See #40847 for follow-up work that would use streaming cursors everywhere, but I don't think it should always be on).

**Comment by J0** [comment]
Hey @itamarst  and @jreback,

Thanks for the input! Will do, will address the comments and revert shortly.

**Comment by J0** [comment]
Yes, I am still interested in working on this and will take a look at this on the weekend

**Comment by simonjayhawkins** [comment]
> Yes, I am still interested in working on this and will take a look at this on the weekend

Thanks @J0 . closing as stale. ping when ready

**Comment by J0** [comment]
Okay, will do. Sorry -- work got a bit busy. Will try to make time for it in this weekend or over an evening.

**Comment by J0** [comment]
Hey! @simonjayhawkins,

I'm working on this right now -- any chance I could get this re-opened?

**Comment by J0** [comment]
Thanks!

**Comment by J0** [comment]
Not too sure what's the best way to go about writing a failing this for this since the main change is in the option passed in

Will think about this but just wanted to check if  @itamarst and @jreback had any suggestions

**Comment by itamarst** [comment]
My guess is the failure is unrelated: it timed out after 60 minutes. I would just push a meaningless change and see if it passes this time.

**Comment by J0** [comment]
Looks like I need approval to run the workflows D: @simonjayhawkins could I trouble you again for help with requesting for approval to run the branch? 

Thank you so much for  -- am aware you probably are quite busy.

**Comment by J0** [comment]
Hey, does anyone have suggestions on how I might get the tests to pass? a little clueless at this point.

**Comment by itamarst** [comment]
In a large project like this, it's possible (a) to have intermittent failures (b) for things to break in ways that are unrelated to your code. So general procedure is:

1. Try rerunning with meaningless commit.
2. If you get the exact same failure, and it's clear that it's not related to your code, merge forward or rebase, depending on how the project does it.

**Comment by J0** [comment]
Workflows :( any chance for help with this again? @simonjayhawkins

File: pandas/io/sql.py [hunk 1]
```diff
@@ -1421,7 +1421,13 @@ def run_transaction(self):
 
     def execute(self, *args, **kwargs):
         """Simple passthrough to SQLAlchemy connectable"""
-        return self.connectable.execution_options().execute(*args, **kwargs)
+        if "chunksize" in kwargs:
```
Comment (jreback) [inline pandas/io/sql.py:1424]: if you are going to do this, then chunksize should be an actual keyword with a default, and a doc-string.

are there *any* tests which actually evaluate this path?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by J0** [comment]
@mroeschke apologies for the delay. I now have time to patch this again -- by any chance could we reopen this? Or would it be better if I opened a new PR? 

Lmk!

**Comment by mroeschke** [comment]
@J0 appears I cannot reopen this pull request as our master branch was renamed to main recently. I would recommend to open a new PR with the change.

**Comment by J0** [comment]
@mroeschke sure, sounds good, will open a new PR! :)

**Comment by J0** [comment]
Re-opened at https://github.com/pandas-dev/pandas/pull/46166

**Pull request closed without merging**